Platform/Authentication/Token Provider/DefaultTokenProvider.swift

```diff
@@ -0,0 +1,228 @@
+    /// supply URL encoded body items which your backend might require to process the request.
+    ///
+    /// If not specified otherwise, the token provider will always add grant_type=client_credentials item
+    /// to the body of the request.
+    public var bodyInjector: BodyInjector?
+    
+    /// An optional logger used by the token provider.
+    public let logger: PPLogger?
+    
+    // MARK: - Initializers
+    
+    /// Create an DefaultTokenProvider which presents either headers, query items or URL encoded body
+    /// items as part of the request. These should be used to identify your application user session to your
+    /// backend so that it can issue a token for the user.
+    ///
+    /// - Parameters:
+    ///     - url: The URL to be called.
+    ///     - headers: An optional dictionary of headers to include in the request. Here you can
+    ///     supply the session or other credientials which your endpoint might require to authenticate
+    ///     the request. If not specified otherwise, the token provider will always set Content-Type header
+    ///     to application/x-www-form-urlencoded.
+    ///     - queryItems: An optional list of query items to include in the request URL. Here you can
+    ///     supply any query items which your endpoint might require to process the request.
+    ///     - body: An optional list of URL encoded body items to include in the request. Here you can
+    ///     supply any body items which your endpoint might require to process the request. If not
+    ///     specified otherwise, the token provider will always add grant_type=client_credentials item
+    ///     to the body of the request.
+    ///     - logger: An optional logger used by the token provider.
+    public init(url: URL, headers: [String : String]? = nil, queryItems: [URLQueryItem]? = nil, body: [URLEncodedBodyItem]? = nil, logger: PPLogger? = nil) {
+        self.url = url
+        self.headers = headers
+        self.queryItems = queryItems
+        self.body = body
+        self.headersInjector = nil
+        self.queryItemsInjector = nil
+        self.bodyInjector = nil
+        self.logger = logger
+    }
+    
+    // MARK: - Token retrieval
+    
+    /// Method called by the SDK to authenticate the user.
+    ///
+    /// - Parameters:
+    ///     - completionHandler: The completion handler that provides
+    ///     `AuthenticationResult` to the SDK.
+    public func fetchToken(completionHandler: @escaping (AuthenticationResult) -> Void) {
+        self.updateRequestParametersIfNeeded()
+        
+        do {
+            let request = try self.prepareRequest()
+            
+            self.fetchToken(with: request, completionHandler: completionHandler)
+        } catch {
+            completionHandler(.failure(error: error))
+        }
+    }
+    
+    // MARK: - Private methods
+    
+    private func updateRequestParametersIfNeeded() {
```

Comment:
I was wondering of the purpose local properties `headers`/`queryItems`/`body`?   If they exist for a good reason then fair enough.  

But if not, you could simplify the class significantly by removing them and this `updateRequestParametersIfNeeded` function as follows (this also makes the `Injector`s all non-optional which is a nice bonus).

```
public var headersInjector: HeadersInjector
public var queryItemsInjector: QueryItemsInjector
public var bodyInjector: BodyInjector

public init(url: URL,
            headers: [String : String]? = nil,
            queryItems: [URLQueryItem]? = nil,
            body: [URLEncodedBodyItem]? = nil,
            logger: PPLogger? = nil) {
    
    self.url = url
    
    self.headersInjector = { headers }
    self.queryItemsInjector = { queryItems }
    self.bodyInjector = { body }

    self.requestFactory = RequestFactory(logger: logger)
    self.logger = logger
}

public func fetchToken(completionHandler: @escaping (AuthenticationResult) -> Void) {
    do {
        let headers = headersInjector()
        let queryItems = queryItemsInjector()
        let body = bodyInjector()
        
        let request = try requestFactory.prepareRequest(url: self.url, headers: headers, queryItems:queryItems, body: body)
        
        Self.fetchToken(with: request, logger: logger, completionHandler: completionHandler)
    } catch {
        completionHandler(.failure(error: error))
    }
}
```

Reply:
Please see the conversation about this requirement on Slack.

Reply:
I think you may have mis-understood my comment here.  On Slack we spoke about the `Injector` properties, here I am discussing the `headers`/`queryItems`/`body` properties.

Reply:
I think that this question should be answered by @mdpye as the above implementation is based on his design.

Reply:
I didn't intend to mandate any internal properties in the sketch I laid out.

I think @OliverPearmain-Pusher's suggestion makes sense in terms of reducing the mutability and number of branches.

Reply:
I removed `header`, `queryItems` and `body` properties.